infrastructure/merklemountainrange/src/merkle_change_tracker.rs

```diff
@@ -0,0 +1,232 @@
+    current_horizon: usize, // how many cp's have you had to date
+    mmr_key: String,
+    object_key: String,
+    init_key: String,
+}
+
+/// This struct is used as a temporary data struct summarizing all changes in a checkpoint.
+/// It saves all the changes made to the MMR as a diff before dumping it to disc. This is done so we can iterate over
+/// changes made to the MMR.
+#[derive(Serialize, Deserialize)]
+pub(crate) struct MerkleCheckPoint {
+    objects_to_add: Vec<ObjectHash>,
+    pub objects_to_del: Vec<ObjectHash>,
+    mmr_to_add: Vec<MerkleNode>,
+}
+impl MerkleCheckPoint {
+    pub(crate) fn add(&mut self, rhs: &mut MerkleCheckPoint) {
+        self.objects_to_add.extend(rhs.objects_to_add.drain(..));
+        self.objects_to_del = Vec::new();
+        self.mmr_to_add.extend(rhs.mmr_to_add.drain(..));
+
+        // The objects to be deleted will now be deleted without record of them as they are older than pruning horizon
+        let mut i = rhs.objects_to_del.len();
+        while i >= 1 {
+            let k = self.objects_to_add.len();
+            if self.objects_to_add[k - 1] == rhs.objects_to_del[i - 1] {
+                self.objects_to_add.remove(i - 1);
+                i -= 1;
+                continue;
+            }
+        }
+    }
+}
+
+impl MerkleChangeTracker {
+    /// create a new change tracker
+    pub fn new() -> MerkleChangeTracker {
+        MerkleChangeTracker {
+            enabled: false,
+            objects_to_save: Vec::new(),
+            objects_to_del: Vec::new(),
+            tree_saved: 0,
+            pruning_horizon: 0,
+            current_horizon: 0,
+            mmr_key: "".to_string(),
+            object_key: "".to_string(),
+            init_key: "".to_string(),
+        }
+    }
+
+    /// initialise the change tracker
+    pub fn init(&mut self, store_prefix: &str, pruning_horizon: usize) {
+        self.enabled = true;
+        self.mmr_key = (store_prefix.clone().to_owned() + &"_mmr_checkpoints".to_string()).to_string();
+        self.object_key = (store_prefix.clone().to_owned() + &"_mmr_objects".to_string()).to_string();
+        self.init_key = (store_prefix.clone().to_owned() + &"_init".to_string()).to_string();
+        self.pruning_horizon = pruning_horizon;
+    }
+
+    /// This function adds a ref to a object to be saved
+    pub fn add_new_data(&mut self, hash: ObjectHash) {
+        if !self.enabled {
+            return;
+        }
+        self.objects_to_save.push(hash.clone());
+    }
+
+    /// This function adds a ref to a object to be saved
+    pub fn remove_data(&mut self, hash: ObjectHash) {
+        if !self.enabled {
+            return;
+        }
+        self.objects_to_del.push(hash.clone());
+    }
+
+    /// Function to save the current checkpoint
+    pub fn save<T, S: MerkleStorage>(
+        &mut self,
+        hashmap: &mut HashMap<ObjectHash, MerkleObject<T>>,
+        mmr: &Vec<MerkleNode>,
```

Comment:
Clippy will say use `&[MerkleNode]` here

Reply:
I will put an ignore flag for clippy here. 
The problem is that I give the controller(the change checker) the pointers to the actual MMR vec and hashmap. The array inside the MMR needs to be a vec,  the only way I can create an array is copy or slice between the two